compose: Add new UI for replying to a message.

Vector73 opened the pull request:
<!-- Describe your pull request here.-->

This PR adds a new way to reply to a message. It adds a new option in the message actions popover "Reply to message" which inserts new reply UI over textarea where the user is composing the message. The reply UI includes:
- User pill of the user who sent the referenced message.
- Link to referenced message
- Reply action buttons: Mention toggle button and remove reply button
- Optionally link to the topic containing the referenced message, if the location of the referenced message is different the reply message.

The main logic to render reply UI for messages exists in `postprocess_content`. If the first block of the rendered content of the message contains both a mention and link to a message, it is converted to a reply block.

The text of the message link is computed by extracting text from the first non reply node of rendered HTML content.

There is a technical issue related to showing topic link in the reply UI for moved messages. If the referenced message has been moved, the topic link is determined by message ID of referenced message. But if the message is not locally fetched, topic link is determined from the link present in message content. This might result in incorrect topic link in reply UI.

Fixes: #36404<!-- Issue link, or clear description.-->

CZO: [#design > message reply feature](https://chat.zulip.org/#narrow/channel/101-design/topic/message.20reply.20feature/with/2330296)

**How changes were tested:**

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details><summary>Screenshots</summary>
<p>

**Dark theme**

<img width="502" height="444" alt="image" src="https://github.com/user-attachments/assets/48bf7ee0-a66b-4dad-a7b3-0e9a6afde4af" />

<img width="1155" height="538" alt="image" src="https://github.com/user-attachments/assets/37892dc4-c609-415b-82bc-27aebd8f00e8" />

<img width="1176" height="523" alt="image" src="https://github.com/user-attachments/assets/216b4e43-c576-4fa2-b99b-f665d580e54f" />

<img width="1158" height="240" alt="image" src="https://github.com/user-attachments/assets/6967d8a4-2a87-46bf-80f3-81b5ec026e7a" />

<img width="1135" height="385" alt="image" src="https://github.com/user-attachments/assets/0da67805-f1cc-42a4-b027-3612accca0fa" />

<img width="1181" height="369" alt="image" src="https://github.com/user-attachments/assets/bd3cd8d6-2417-453d-a78e-e691992f171c" />

**Light theme**

<img width="404" height="515" alt="image" src="https://github.com/user-attachments/assets/cc7399cf-40ad-4f97-ab3d-75243f80c9bc" />

<img width="1139" height="572" alt="image" src="https://github.com/user-attachments/assets/c3fcc726-6037-4d10-9126-1b3804f02ab9" />

<img width="1164" height="214" alt="image" src="https://github.com/user-attachments/assets/7cc3335e-fe0c-4764-ba4a-16ad978db14d" />

<img width="1167" height="442" alt="image" src="https://github.com/user-attachments/assets/6c644a17-cb37-40b7-bae0-b688d37fc15b" />



</p>
</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
I'm not sure the message link in the compose box should be clickable. It's kind of confusing if you do end up clicking it.

Vector73 committed:
compose: Add new UI for replying to a message.

This PR adds a new way to reply to a message. It adds a new option in
the message actions popover "Reply to message" which inserts new
reply UI over textarea where the user is composing the message.

Fixes #36404.

timabbott commented on web/src/click_handlers.ts:
```diff
@@ -479,6 +478,41 @@ export function initialize(): void {
         sidebar_ui.hide_userlist_sidebar();
     });
 
+    $("body").on("click", ".mention-button", function (e) {
```
`mention-button` is too generic of a selector name. Maybe `compose-reply-mention-toggle` would do?

benlubas commented:
@alya, if I take this over can it move forward?

@Vector73, are you still interested in this feature?

laurynmm commented:
Closing this in favor of #38784, which is an updated version of these changes with this review feedback implemented: https://github.com/zulip/zulip/pull/36956#discussion_r2714312917.
